### ISSUE TITLE
Include least squares parameter optimization

### ISSUE THREAD

**Issue by mrp089** [SimVascular/svZeroDSolver#24]
Edit: See comment below, this is a **non-linear** least-squares problem.

**TL;DR**: Implement a (linear) least squares fit of all parameters (`R,C,L,stenosis,...`) in a 0D model in `svZeroDPlus` from given 3D (or other) measurements.

The 0D residual (assuming no time-dependency in parameters) is $$\mathbf{R} (\boldsymbol{\alpha}, \mathbf{y}, \dot{\mathbf{y}}) = \mathbf{E} (\boldsymbol{\alpha},\mathbf{y}) \cdot \dot{\mathbf{y}} + \mathbf{F} (\boldsymbol{\alpha},\mathbf{y})  \cdot \mathbf{y} + \mathbf{c} (\boldsymbol{\alpha}, \mathbf{y})$$ with solution vector $\mathbf{y} \in \mathbb{R}^N$ (flow, pressure), LPN parameters $\boldsymbol{\alpha} \in \mathbb{R}^P$, system matrices $\mathbf{E}, \mathbf{F} \in \mathbb{R}^{N\times N}$, system vector $\mathbf{c} \in \mathbb{R}^N$.

The (standard) 0D initial-boundary value problem that `svZeroDPlus` solves is $$\text{Solve} \quad \mathbf{R} (\boldsymbol{\alpha}, \mathbf{y}, \dot{\mathbf{y}}) = \mathbf{0} \quad \text{for~} \mathbf{y}~ \forall\,t$$ with given parameters $\boldsymbol{\alpha}$.

This residual can be repurposed, formulating the 0D-3D calibration as a least squares problem: $$\min_{\boldsymbol{\alpha}} S, \quad\text{with~} S = \sum_i^{D} R_i^2(\boldsymbol{\alpha}, y_i^\text{3D}, \dot{y}_i^\text{3D} )$$ with given 3D solution $\mathbf{y}^\text{3D}, \dot{\mathbf{y}}^\text{3D}$ at all data points $D=N\times T.$ Note that this allows us to arbitrarily combine observations. We don't necessarily need to supply an observation vector for the whole system dimension $N$ but could also only supply a subset (e.g. at the outlets). However, it is never guaranteed that there's a unique solution, it's always a least-squares fit.

Fortunately, the residual $\mathbf{R}$ depends linearly on parameters $\boldsymbol{\alpha}$ (note that this is different than the dependency of the residual on the solution vector $\mathbf{y}$):
$$\mathbf{R} (c\boldsymbol{\alpha} +\boldsymbol{\beta}, \mathbf{y}^\text{3D}, \dot{\mathbf{y}}^\text{3D}) = c\mathbf{R} (\boldsymbol{\alpha}, \mathbf{y}^\text{3D}, \dot{\mathbf{y}}^\text{3D}) + \mathbf{R} (\boldsymbol{\beta}, \mathbf{y}^\text{3D}, \dot{\mathbf{y}}^\text{3D})$$

Thus, we can rewrite the residual as a linear combination $$\mathbf{r} (\boldsymbol{\alpha}, \mathbf{y}^\text{3D}, \dot{\mathbf{y}}^\text{3D}) = \mathbf{Y} (\mathbf{y}^\text{3D}, \dot{\mathbf{y}}^\text{3D}) - \mathbf{X}(\mathbf{y}^\text{3D}, \dot{\mathbf{y}}^\text{3D}) \cdot \boldsymbol{\alpha}$$ with space-time residual $\mathbf{r} \in \mathbb{R}^{D}$,  vector $\mathbf{Y} \in \mathbb{R}^{D}$ and matrix $\mathbf{X} \in \mathbb{R}^{D \times P}$.

We can then find the optimal parameters $\hat{\boldsymbol\alpha}$ by solving the [linear least squares problem](https://en.wikipedia.org/wiki/Linear_least_squares)
$$\hat{\boldsymbol\alpha} = \left( \mathbf{X}^\text{T}\mathbf{X} \right)^{-1} \cdot \mathbf{X}^\text{T} \cdot \mathbf{Y}$$ This problem is computationally easy to solve since $\mathbf{X}^\text{T}\mathbf{X} \in \mathbb{R}^{P\times P}$.

The matrices $\mathbf{X},\mathbf{Y}$ can be obtained from linearizing the residual w.r.t. parameters:
$$- \mathbf{X} = \frac{\partial \mathbf{r}}{\partial \boldsymbol{\alpha}} = \frac{\partial \mathbf{E}}{\partial \boldsymbol{\alpha}} \cdot \dot{\mathbf{y}}^\text{3D} + \frac{\partial \mathbf{F}}{\partial \boldsymbol{\alpha}} \cdot {\mathbf{y}}^\text{3D} + \frac{\partial \mathbf{c}}{\partial \boldsymbol{\alpha}}$$ $$\mathbf{Y} = \text{all terms in~} \mathbf{E}, \mathbf{F}, \mathbf{c}~ \text{that do not depend on}~ \boldsymbol\alpha$$ Note that the derivatives $\partial/\partial\boldsymbol\alpha$ are super easy to compute and are mostly zero matrices with $\pm1$ where `R,C,L` are defined. Only the derivative w.r.t to the stenosis factor depends on $\mathbf{y}$ via the flow rate.

I think the implementation steps in `svZeroDPlus` are:
- Read in space-time measurements $\mathbf{y}^\text{3D}$ from file ($\dot{\mathbf{y}}^\text{3D}$ can be computed internally consistent with gen $\alpha$ time integration)
For now, we can assume that the measurements have the same dimension/ordering as the solution vector. In later iterations, we can also allow spatially defining the solution (e.g. only at the outlets).
- Define global parameter vector $\boldsymbol\alpha$
- Derive $\partial\mathbf{r}/\partial\boldsymbol{\alpha}$ for every 0D element
- Calculate $\mathbf{X}, \mathbf{Y}$ on every 0D element
- Assemble global $\mathbf{X}, \mathbf{Y}$
- Solve global optimization problem for $\hat{\boldsymbol\alpha}$
- Output residual and 0D input file with optimal parameters $\hat{\boldsymbol\alpha}$

For testing, we could try to fit the parameters of an existing 0D solution and see if we obtain the ground truth. This could also be a cool exercise to see if we can still estimate the parameters from sparse data (e.g. only at the outlets) and study how well that solution generalizes to e.g. different boundary conditions.

@menon-karthik and @richterjakob expressed initial interest to implement such an optimization. Please coordinate in this issue who does what.

Update 2023-03-09: Fixed computation of $\mathbf{Y}$

**Comment by mrp089** [comment]
This approach can also be seen as a linear regression analysis, as described in [pant14](https://doi.org/10.1002/cnm.2692) to calibrate 0D parameters from a 3D solution. Although, I think this approach is applied locally for every 0D element which should be similar to our previous local tuning approaches.

**Comment by mrp089** [comment]
Following up on a discussion with @richterjakob on the treatment of boundary conditions.

**Dirichlet boundary conditions** (DBCs)
E.g., `FlowReferenceBC` or `PressureReferenceBC`. These blocks don't depend on the parameter vector $\boldsymbol\alpha$, thus adding a zero row globally to $\mathbf{X}$. Their degrees of freedom could be removed from $\mathbf{y}^\text{3D},\dot{\mathbf{y}}^\text{3D}$ and subsequently from $\mathbf{r}$. However, it might be easier to implement (a trivial) $\mathbf{Y}$ for them and assemble the DBCs like all other blocks. The zero row in $\mathbf{X}$ would make sure that the observations in $\mathbf{Y}$ do not contribute to the residual $\mathbf{r}$.

**LPN boundary conditions** (LBCs)
E.g., `WindkesselBC` or `OpenLoopCoronaryBC`. Their parameters should be excluded from the parameter vector $\boldsymbol\alpha$ and subsequently from $\mathbf{X},\mathbf{Y}$ when assembling the least-squares system since we don't want to estimate them.

**Comment by mrp089** [comment]
After some investigation by @richterjakob it turned out that the assumption that the residual $\mathbf{R}$ depends linearly on the parameters $\boldsymbol\alpha$ was wrong.

Currently, the way the `BloodVessel` block [is implemented](https://stanfordcbcl.github.io/svZeroDPlus/cpp/class_m_o_d_e_l_1_1_blood_vessel.html) one might think that it depends linearly on `R`, `C`, `L`, `sten`. However, this formulation requires an internal variable `P_c` that can only be computed with the knowledge of $\boldsymbol\alpha$. If formulated without this internal variable (reducing the size of the block from 3 to 2), then a product $R\cdot C$ appears.

We can thus not solve for the optimal parameters $\hat{\boldsymbol\alpha}$ in a linear least-squares problem. However, we can use what @richterjakob has implemented so far and use it to solve a Gauss-Newton problem to iteratively determine $\hat{\boldsymbol\alpha}$: $$\boldsymbol\alpha^{s+1} = \boldsymbol \alpha^s - (\mathbf{J}^\text{T} \mathbf{J} )^{-1} \cdot \mathbf{J}^\text{T} \cdot \mathbf{r}(\boldsymbol \alpha^s)$$ where the Jacobian matrix $\mathbf{J}$ is equal to $\mathbf{X}$ in the linear least-squares problem.

**Comment by mrp089** [comment]
Future development ideas that go beyond #31:
- calculate derivative `dy` of solution vector `y` inside `svZeroDPlus` so it doesn't have to be user-supplied
- use the `svzerodsolver` output file `.csv` as input for `svzerodcalibrator`
